llm_studio/src/utils/config_utils.py

```diff
@@ -61,3 +78,152 @@ def load_config(config_path: str, config_name: str = "Config"):
+
+    return None
+
+
+def parse_cfg_dataclass(cfg: ConfigProblemBase) -> List[Dict]:
+    """Returns all single config settings for a given configuration
+
+    Args:
+        cfg: configuration
+    """
+
+    items = []
+
+    parent_element = get_parent_element(cfg)
+    if parent_element:
+        items.append(parent_element)
+
+    cfg_dict = cfg.__dict__
+    type_annotations = cfg.get_annotations()
+    cfg_dict = {key: cfg_dict[key] for key in cfg._get_order()}
+
+    for k, v in cfg_dict.items():
+
+        if k.startswith("_") or cfg._get_visibility(k) < 0:
+            continue
+
+        if any([x in k for x in ["api"]]):
+            continue
+
+        type_annotation = type_annotations[k]
+
+        if type_annotation in KNOWN_TYPE_ANNOTATIONS:
+            if type_annotation == float:
+                v = float(v)
+            t = [{k: v}]
+        elif dataclasses.is_dataclass(v):
+            elements_group = parse_cfg_dataclass(cfg=v)
+            t = elements_group
+        else:
+            raise _get_type_annotation_error(v, type_annotations[k])
+
+        items += t
+
+    return items
+
+
+def save_config_yaml(path: str, cfg: ConfigProblemBase) -> None:
+    """Saves config as dill file
+
+    Args:
+        path: path of file to save to
+        cfg: config to save
+    """
+    cfg_dict = convert_cfg_to_nested_dictionary(cfg)
+    cfg_dict["experiment_name"] = cfg.experiment_name
+    cfg_dict["output_directory"] = cfg.output_directory
+    cfg_dict["llm_backbone"] = cfg.llm_backbone
+    with open(path, "w") as fp:
+        yaml.dump(cfg_dict, fp, indent=4)
+
+
+def load_config_yaml(path: str) -> ConfigProblemBase:
+    """Loads config from yaml file
+
+    Args:
+        path: path of file to load from
+    Returns:
+        config object
+    """
+    with open(path, "r") as fp:
+        cfg_dict = yaml.load(fp, Loader=yaml.FullLoader)
+
+    cfg = ConfigProblemBase(
+        output_directory=cfg_dict.get("output_directory", ConfigProblemBase.output_directory),
+        experiment_name=cfg_dict.get("experiment_name", generate_experiment_name()),
+        llm_backbone=cfg_dict.get("llm_backbone", ConfigProblemBase.llm_backbone),
+        dataset=ConfigNLPCausalLMDataset.from_dict(cfg_dict.get("dataset", {})),
+        tokenizer=ConfigNLPCausalLMTokenizer.from_dict(cfg_dict.get("tokenizer", {})),
+        augmentation=ConfigNLPAugmentation.from_dict(cfg_dict.get("augmentation", {})),
+        architecture=ConfigNLPCausalLMArchitecture.from_dict(
+            cfg_dict.get("architecture", {})
+        ),
+        training=ConfigNLPCausalLMTraining.from_dict(cfg_dict.get("training", {})),
+        prediction=ConfigNLPCausalLMPrediction.from_dict(
+            cfg_dict.get("prediction", {})
+        ),
+        environment=ConfigNLPCausalLMEnvironment.from_dict(
+            cfg_dict.get("environment", {})
+        ),
+        logging=ConfigNLPCausalLMLogging.from_dict(cfg_dict.get("logging", {})),
+    )
```

Comment:
any way to have that dynamically grab the problem type? We will need to consider if we add other configs. Likely need to add to the yaml, too.

Reply:
I added `cfg_dict["problem_type"] = "text_causal_language_modeling"` when saving the yaml.
Dynamic mapping to different problem types should thus be possible. `load_config_yaml` can be generalized once we decide to add additional configurations.